### PR TITLE
Introduce context struct

### LINKED ISSUE #30: Introduce custom context types
All functions in sylvia take some context on the first value - for now, those are typically tuples of cosmwasm-std types, but it makes using them weird. I propose to define context type for all kinds of messages, for example:
```rust
struct ExecContext {
  deps: DepsMut,
  env: Env,
  info: MessageInfo,
}
```
So it is a bit less noise. It is very easy to implement, as sylvia always calls `.into()` on passed tuples, so all those context structures need is to implement `From<...>` trait.

### PR DESCRIPTION
**Pull request by jawoznia**
close #30

### CODE REVIEW

File: sylvia-derive/src/parser.rs [hunk 1]
```diff
@@ -109,15 +109,20 @@ impl MsgType {
...
+                #sylvia ::types::InstantiateCtx
             },
             Migrate => quote! {
-                (cosmwasm_std::DepsMut, cosmwasm_std::Env)
+                #sylvia ::types::MigrateCtx,
             },
             Query => quote! {
-                (cosmwasm_std::Deps, cosmwasm_std::Env)
+                #sylvia ::types::QueryCtx
             },
```
Comment (hashedone) [inline sylvia-derive/src/parser.rs:126]: Here I am not exactly sure - we obviously want to support the new context types, but supporting original tuples or custom context types would also be nice. Iw ould prefer here to emit something like:

```
impl From<(#sylvia ::cw_std::DepsMut, #sylvia ::cw_std::Env, #sylvia ::cw_std::MessageInfo)>
```

I am not sure, but I think we use this in function arguments position, and this way you will be able to use any type constructible from tuple context. Obviously some `.into()` in generated could would have to be added.

**Review by hashedone — COMMENTED** [review]
We could consider how we emit signatures of functions taking context types.

File: sylvia-derive/src/parser.rs (hunk 1 above)
Comment (jawoznia) [reply sylvia-derive/src/parser.rs]: reverted to simple tuple

File: contracts/cw1-subkeys/src/contract.rs [hunk 2]
```diff
@@ -43,62 +44,63 @@ impl Cw1SubkeysContract<'_> {
...
-                .unwrap_or_default();
-
-            let mut allowance = allow
-                .filter(|allow| !allow.expires.is_expired(&env.block))
-                .unwrap_or_default();
+        let spender = ctx.deps.api.addr_validate(&spender)?;
+        ensure_ne!(
+            ctx.info.sender,
+            spender,
+            ContractError::CannotSetOwnAccount {}
```
Comment (hashedone) [inline contracts/cw1-subkeys/src/contract.rs:73]: Why use `{}` for errors? They don't need it; they are not serialized.

**Review by hashedone — APPROVED** [review]
LGTM

File: contracts/cw1-subkeys/src/contract.rs (hunk 2 above)
Comment (jawoznia) [reply contracts/cw1-subkeys/src/contract.rs:73]: Agree. This is unnecessary

Comment (jawoznia) [reply contracts/cw1-subkeys/src/contract.rs:73]: Will deliver fix in separate PR

Comment (jawoznia) [reply contracts/cw1-subkeys/src/contract.rs:73]: Removal of curly braces on errors #95

**Pull request merged**